simulation/halsim_ws_server/src/test/native/cpp/main.cpp

```diff
@@ -0,0 +1,125 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "gtest/gtest.h"
+#include <hal/Main.h>
+#include <hal/HALBase.h>
+#include <hal/simulation/DIOData.h>
+#include "HALSimWSServer.h"
+#include "WebServerClientTest.h"
+#include <wpi/uv/Loop.h>
+#include <iostream>
+
+using namespace wpilibws;
+
+extern "C" int HALSIM_InitExtension(void); // in simulation/halsim_ws_server/src/main/native/cpp/main.cpp
+
+class WebServerIntegrationTest : public ::testing::Test {
+public:
+  WebServerIntegrationTest() {
+    // Needed to initialize providers and webserver loop
+    HALSIM_InitExtension();
+
+    m_webserver_client = std::shared_ptr<WebServerClientTest>((new WebServerClientTest()));
+    WebServerClientTest::SetInstance(m_webserver_client);
+    auto webserver = HALSimWeb::GetInstance();
+    auto m_loop = webserver->GetLoop();
```

Comment:
This is the part I'm most uncomfortable with.. feels just a tad bit dirty

Reply:
We discussed offline. Changing this workflow would mean we'd have to change providers from being a reference to being a pointer. To avoid large refactoring, we will leave it as is for now.